libraries/chain/webassembly/crypto.cpp

```diff
@@ -250,4 +251,141 @@ namespace eosio { namespace chain { namespace webassembly {
+      return return_code::success;
+   }
+
+   int32_t interface::bls_g1_exp(span<const char> points, span<const char> scalars, const uint32_t n, span<char> result) const
+   {
+      if(points.size() != n*144 ||  scalars.size() != n*32 ||  result.size() != 144)
+         return return_code::failure;
+      std::vector<bls12_381::g1> pv;
+      std::vector<std::array<uint64_t, 4>> sv;
+      pv.reserve(n);
+      sv.reserve(n);
+      for(uint32_t i = 0; i < n; i++)
+      {
+         bls12_381::g1 p = bls12_381::g1::fromJacobianBytesLE({reinterpret_cast<const uint8_t*>(points.data() + i*144), 144}, false, true);
+         std::array<uint64_t, 4> s = bls12_381::scalar::fromBytesLE<4>({reinterpret_cast<const uint8_t*>(scalars.data() + i*32), 32});
+         pv.push_back(p);
+         sv.push_back(s);
+         if(i%10 == 0)
+            context.trx_context.checktime();
+      }
+      bls12_381::g1 r = bls12_381::g1::multiExp(pv, sv);
+      r.toJacobianBytesLE({reinterpret_cast<uint8_t*>(result.data()), 144}, true);
+      return return_code::success;
+   }
+
+   int32_t interface::bls_g2_exp(span<const char> points, span<const char> scalars, const uint32_t n, span<char> result) const
+   {
+      if(points.size() != n*288 ||  scalars.size() != n*32 ||  result.size() != 288)
+         return return_code::failure;
+      std::vector<bls12_381::g2> pv;
+      std::vector<std::array<uint64_t, 4>> sv;
+      pv.reserve(n);
+      sv.reserve(n);
+      for(uint32_t i = 0; i < n; i++)
+      {
+         bls12_381::g2 p = bls12_381::g2::fromJacobianBytesLE({reinterpret_cast<const uint8_t*>(points.data() + i*288), 288}, false, true);
+         std::array<uint64_t, 4> s = bls12_381::scalar::fromBytesLE<4>({reinterpret_cast<const uint8_t*>(scalars.data() + i*32), 32});
+         pv.push_back(p);
+         sv.push_back(s);
+         if(i%6 == 0)
+            context.trx_context.checktime();
+      }
+      bls12_381::g2 r = bls12_381::g2::multiExp(pv, sv);
+      r.toJacobianBytesLE({reinterpret_cast<uint8_t*>(result.data()), 288}, true);
+      return return_code::success;
+   }
+
+   int32_t interface::bls_pairing(span<const char> g1_points, span<const char> g2_points, const uint32_t n, span<char> result) const
+   {
+      if(g1_points.size() != n*144 ||  g2_points.size() != n*288 ||  result.size() != 576)
+         return return_code::failure;
+      std::vector<std::tuple<bls12_381::g1, bls12_381::g2>> v;
+      v.reserve(n);
+      for(uint32_t i = 0; i < n; i++)
+      {
+         bls12_381::g1 p_g1 = bls12_381::g1::fromJacobianBytesLE({reinterpret_cast<const uint8_t*>(g1_points.data() + i*144), 144}, false, true);
+         bls12_381::g2 p_g2 = bls12_381::g2::fromJacobianBytesLE({reinterpret_cast<const uint8_t*>(g2_points.data() + i*288), 288}, false, true);
+         bls12_381::pairing::add_pair(v, p_g1, p_g2);
+         if(i%4 == 0)
+            context.trx_context.checktime();
+      }
+      bls12_381::fp12 r = bls12_381::pairing::calculate(v);
```

Comment:
In the case of `v.size() == 80000` how long does `bls12_381::pairing::calculate()` take?

Reply:
On my machine about 20 seconds. I was thinking about it after I read you previous comment: Does that mean, we need to be able to pass the `yield()` function to `pairing::calculate()` and execute it there in reasonable intervals as well?

Reply:
> since now with the `checktime()` inside the loop it wouldn't be possible to add that many pairs to `v` anymore

yeah, good point. But is there some number that takes, for example, 25ms to make it through the loop, but then takes 100ms to make it through `bls12_381::pairing::calculate()`? That too would be considered a security defect. I'm not sure of the timing for these operations so maybe it's a non issue.

Reply:
Good question. So I did some measuring again:

It takes `v.size() == 58` pairs in order for the entire loop to take `25ms` (on my machine). The call to `pairing::calculate(v)` then takes about an additional `14.7ms`.

I'm not exactly sure how those number would vary on a faster machine but I guess it would be a similar ratio between loop execution and paring calculation.

Reply:
It's "only" ~10ms but I think we should be prudent and protect against that -- it definitely would fail previous guidelines we had back in the bug bounty days.

Since we control the library, the best choice is to pass in the ability for `bls12_381::pairing::calculate()` to poll a yield function. An example of that is,
https://github.com/AntelopeIO/leap/blob/7c0694b4c965bf12078680cb4062f092554be0c5/libraries/chain/webassembly/crypto.cpp#L137-L139
where the declaration of that is
```c++
/// pairing_check calculates the Optimal Ate pairing for a set of points.
///  @param marshaled_g1g2_pairs marshaled g1 g2 pair sequence
///  @return -1 for unmarshal error, 0 for unsuccessful pairing and 1 for successful pairing
int32_t pairing_check(std::span<const uint8_t> marshaled_g1g2_pairs, std::function<void()> yield);
```
if you wanted to make the yield optional for users, you could potentially do something like
```c++
int32_t pairing_check(std::span<const uint8_t> marshaled_g1g2_pairs, std::function<void()> yield = std::function<void()>());
```
or just require users to pass in a stub function. Either way is fine; importantly either way ultimately still allows the library to used outside of leap.

(there is another option: add a subjective limit to the size of `n`; for example
https://github.com/AntelopeIO/leap/blob/7c0694b4c965bf12078680cb4062f092554be0c5/libraries/chain/webassembly/crypto.cpp#L146-L156
but I can't find any compelling reason to pick this choice due to our control of the bls lib)

Reply:
Agree. Since there is a `leap` branch of the library now, let's try to tailor it perfectly for leap integration.

I added a `yield` function pointer parameter to `pairing::calculate()` (`pairing::miller_loop()` respectively) and `g2::multiExp()` since this one could also run for about `15ms` in the `25ms`-loop worst case. Here are the numbers for all three functions (my machine):

- G1 multi-exponentiation: with `v.size() == 161` the loop takes `25ms` => `g1::multiexp(v)` then takes only `5.1ms`
- G2 multi-exponentiation: with `v.size() == 96` the loop takes `25ms` => `g2::multiexp(v)` then takes `15.3ms`
- Pairing: with `v.size() == 58` the loop takes `25ms` => `pairing::calculate(v)` then takes `14.7ms`

The timings of `g1::multiExp()` should be fine, I guess, since `25ms + 5.1ms ≅ 30ms`. So if `5ms` to `7ms` is an acceptable 'overhead' for the execution time of either `pairing::calculate()` or `g1/g2::multiExp()` after their corresponding loops (assuming the worst case vector sizes listed above with `25ms` preceding loop execution time) then only one `yield()` call approximately in the middle of execution of `pairing` and `multiexp` should be fine, right? Worst case scenario is then about `32ms` total execution time in case the transaction fails.

It was a bit tricky though, to integrate the `yield()` call accurately into the middle of the `pairing` and `mutliexp` functions because there are multiple loops (including nested loops) of which some depend on `v.size()`. However, I believe I found a good way to add it approximately in the middle of each function's execution path, so that the time check is performed once if `v` exceeds a certain size.

Reply:
@mschoenebeck: I think we should pass the yield function into `g1::multiExp()` as well for consistency.

We should not assume that there will always be a limit of 30 ms in a transaction. We are proposing that the network upgrade to a higher max transaction time. With a max transaction time of 150 ms, the user could pass a 5x larger array and thus increase the time of the `g1::multiexp` function call by 5x as well.